x/programs/rust/examples/counter/src/lib.rs

```diff
@@ -70,34 +71,36 @@ mod tests {
         let owner_key = String::from("owner");
         let alice_key = Param::Key(Key::Ed25519(String::from("alice")));
 
-        let mut plan = Plan::new(&owner_key);
-
-        plan.add_step(Step::create_key(Key::Ed25519(owner_key.clone())));
-
-        plan.add_step(Step {
-            endpoint: Endpoint::Key,
-            method: "key_create".into(),
-            params: vec![alice_key.clone()],
-            max_units: 0,
-        });
-
-        plan.add_step(Step {
-            endpoint: Endpoint::Execute,
-            method: "program_create".into(),
-            max_units: 1000000,
-            params: vec![Param::String(PROGRAM_PATH.into())],
-        });
-
-        let plan_responses = simulator.run_plan(plan).unwrap();
-
-        assert!(
-            plan_responses.iter().all(|resp| resp.base.error.is_none()),
-            "error: {:?}",
-            plan_responses
-                .iter()
-                .filter_map(|resp| resp.base.error.as_ref())
-                .next()
-        );
+        simulator
+            .run_step::<()>(
+                &owner_key,
+                &Step::create_key(Key::Ed25519(owner_key.clone())),
+            )
+            .unwrap();
+
+        simulator
+            .run_step::<()>(
+                &owner_key,
+                &Step {
+                    endpoint: Endpoint::Key,
+                    method: "key_create".into(),
+                    params: vec![alice_key.clone()],
+                    max_units: 0,
+                },
+            )
+            .unwrap();
+
+        simulator
+            .run_step::<()>(
+                &owner_key,
+                &Step {
+                    endpoint: Endpoint::Execute,
+                    method: "program_create".into(),
+                    max_units: 1000000,
+                    params: vec![Param::String(PROGRAM_PATH.into())],
+                },
+            )
+            .unwrap();
```

Comment:
If you do what I suggest with the `repsonse` method, you wouldn't need to pass in the type argument to `run_step` and that would clean up this code. 

Also, a test without a single assertion isn't a great test (the test before wasn't great, but it's more apparent now). 

Please make the test function return a `Result<(), Error>` so that we actually have good error messages.

Reply:
I miss Python's `cached_property` so much...

Reply:
but I implemented your suggestion, good catch thanks!